File: Data Structures & Algorithms/trapping-rain-water/submission-0.java
class Solution {
    public int trap(int[] height) {
        int start = 0;
        int end = height.length - 1;

        int maxLeft = height[start];
        int maxRight = height[end];
        int total = 0;

        while (start <= end) {
            if (maxLeft < maxRight) {
                total += Math.max(0, (Math.min(maxLeft, maxRight) - height[start]));
                maxLeft = Math.max(height[start], maxLeft);
                start++;
            }
            else {
                total += Math.max(0, (Math.min(maxLeft, maxRight) - height[end]));
                maxRight = Math.max(height[end], maxRight);
                end--;
            }
        }
        return total;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        start, end = 0, 0
        res = 0
        count = {}
        
        while (end < len(s)):
            count[s[end]] = 1 + count.get(s[end], 0)
            if (end - start + 1 - max(count.values()) > k):
                count[s[start]] -= 1
                start+=1    
            res = max(res, end - start + 1)
            end+=1
        return res